### ISSUE TITLE
Ability to get strongref'd heapptr 

### ISSUE THREAD

**Issue by fatcerberus** [svaarala/duktape#176]
Currently the pointer returned by duk_get_heapptr() is a borrowed, weak reference--the object can still be garbage collected.  Would there be any possibility of allowing strong references for these?  This would allow me to, e.g. save a pointer to a function passed in from JS code to be called later without having to put it into the global stash as I do currently.

**Comment by svaarala** [comment]
Yes, there's been discussion about adding "pin" and "unpin" calls. Technically they'd:

- Bump reference count (unbump when unpin)
- Flag the object so that mark-and-sweep won't sweep it

The mark-and-sweep flagging needs some thought re: stacked pinning; there'd need to be a pin count (otherwise, when to unflag?). This is particularly an issue if Duktape is built with reference counting disabled.

**Comment by fatcerberus** [comment]
Wouldn't the pin count have to maintained separately from the refcount anyway? (otherwise unmatched pins/unpins could knock the refcount out of sync, leading to, e.g. early deallocation)  So it seems like it shouldn't be an issue even with refcounting disabled.  You flag (and incref, if applicable) on the first pin, unflag and decref on the last unpin.

That's how I would implement it, at least.

**Comment by svaarala** [comment]
Unbalanced pin/unpin operations can't really work anyway - they'll lead to either memory leaks or, again, early deallocation (if there are no actual refcounts and there is one unpin too many). So I don't really see how that would work?

I'd want to avoid another header field because it'd need to be a count field which is expensive on low memory environments. There's been a lot of effort into minimizing header field sizes for these environments.

**Comment by svaarala** [comment]
One option would be to implement pinning simply as a shorthand for storing the object together with a pin count in the heap stash somewhere. This would make the pin count external and avoid increasing the header size for all objects, and would work well if the majority of objects was not pinned.

**Comment by fatcerberus** [comment]
That might work.  This is basically what I do now if I want to hold onto a heapptr - I stash the object.  It's just that it's a lot of extra boilerplate code whose purpose isn't immediately obvious to someone reading the source.  So having Duktape do it for me would be awesome.

As for the early deallocation, I'm okay with it happening as long as it only affects C code holding a heapptr.  The thing that concerns me is the possibility of an object being freed while it's still reachable from ecma code - the ensuing segfaults would be a nightmare to debug.

**Comment by fatcerberus** [comment]
Is it too late to get this into 1.4.0?  I was considering opening a pull request to implement this using the stash method discussed above.

**Comment by JoshEngebretson** [comment]
Being able to pin/unpin a heapptr would be welcome here, it would greatly simplify book-keeping code which currently stashes objects to keep them alive

**Comment by svaarala** [comment]
It's probably too close for 1.4.0 - there are a few approaches (stash helper, pinning) and it'd be nice to avoid overlapping mechanisms if possible.

**Comment by fatcerberus** [comment]
I did an experiment with implementing a stash helper, it ended up being very complicated.  The algorithm was something like this:

1. Let `ptr` be the heap pointer to be pinned.
2. Let `stash` be the global stash.
3. Let `key` be a string derived from `ptr`, e.g. `DEADBEEF`
4. Let `entry` be `stash._Pinned[key]`
5. If `entry.pinCount` is undefined, set it to 1.
6. Otherwise, increment `entry.pinCount`.
7. Set `entry.object` to the object pointed to by `ptr`.

Reverse the process for an unpin, deleting `stash._Pinned[key]` if the pin count reaches 0.  This algorithm is simple enough in theory, but gets very confusing to follow using the value stack API (particularly with the intermediate steps to create the necessary objects if they don't exist).

What are some alternative methods?

**Comment by fatcerberus** [comment]
@svaarala I forgot about this issue until just now, mainly because I found myself having to implement  another stashing solution. :)  Once again I needed a way to hold on to an Ecmascript function to be called at some arbitrary point in the future, possibly more than once.  About halfway through the implementation I decided the manual stash management was too clunky and came up with these generic pinning functions instead:
https://github.com/fatcerberus/minisphere/blob/3beaa4a1287eb7ee8b08bb3c224ed2c8d9e54412/src/compiler/utility.c#L4-L79

Complete with reference counting. :)

Would that be something worth adding as an example in the Duktape repo?

**Comment by kungfooman** [comment]
@fatcerberus Thanks for that, really simple to use it. I just didn't know where to find the missing function, so I searched it (other people probably wonder the same): https://github.com/fatcerberus/minisphere/blob/3beaa4a1287eb7ee8b08bb3c224ed2c8d9e54412/src/shared/duk_rubber.c#L31

	duk_idx_t
	dukrub_push_bare_object(duk_context* ctx)
	{
	#if DUK_VERSION >= 19999
		return duk_push_bare_object(ctx);
	#else
		duk_idx_t idx;

		idx = duk_push_object(ctx);
		duk_push_undefined(ctx);
		duk_set_prototype(ctx, -2);
		return idx;
	#endif
	}

I like to use heapptrs with this reference counting, since I can now "glue together" the JavaScript objects with my C++ objects. Even just tested to set a finalizer via Duktape.fin and it works nicely (called after some seconds).

**Comment by fatcerberus** [comment]
Glad to hear you found it useful. :)

In general I've found it to be a lot easier just to pass a pointer around rather than having to (manually) stash the object and pull it out later.  Avoids the need to cross-reference other parts of the code to find out the exact place you stashed something.

**Comment by fatcerberus** [comment]
@svaarala Forgot that I posted this all those years ago :)

Regarding your aversion to pinning mentioned in https://github.com/svaarala/duktape/issues/1015#issuecomment-323816136: I do agree Duktape's value stack discipline makes it a lot more convenient than engines that enforce explicit rooting/pinning, because to some extent the value stack is "embedded" within the C stack.  However, that doesn't discount use cases where explicit pinning, either through the stash or some hypothetical pinning API, is the only viable solution to a problem--for example, any time you need to store a strong reference in the (native) heap.  This comes up often in applications like miniSphere, where the boundary between native and Ecmascript code is quite blurred and objects get passed freely between the two domains.

I do agree with you that any solution in this direction would need some way to look at active pins, to aid debugging memory leaks and such.  To that end, perhaps what you suggested above would be best:
> One option would be to implement pinning simply as a shorthand for storing the object together with a pin count in the heap stash somewhere.

**Comment by svaarala** [comment]
A simple convenience API with no internal changes for pinning could use the raw heap pointer as a key, and then maintain a counter with that key. A pointer as a key gets string coerced which is not ideal of course, but the key is unique and stable, and just having the pointer is enough to unpin.

As an example (untested):

```c
/* Pin value on top of stack and pop the value. Return duk_get_heapptr() for the value as a convenience. */
void *duk_pin_value(duk_context *ctx) {
    void *ptr = duk_get_heapptr(ctx, -1);
    duk_uint_t pin_count;

    if (ptr == NULL) {
        /* Attempt to pin a non-heap-object is a nop. */
        duk_pop(ctx);
        return NULL;
    }

    duk_push_heap_stash(ctx);
    duk_push_pointer(ctx, ptr);  /* [ ... val stash ptr ] */
    (void) duk_get_prop(ctx, -2);
    pin_count = duk_to_uint(ctx, -1);  /* 0 if property missing */
    duk_push_pointer(ctx, ptr);
    duk_push_uint(ctx, pin_count + 1);  /* -> [ ... val stash oldcount ptr newcount ] */
    duk_put_prop(ctx, -4);  /* stash[ptr]++; -> [ ... val stash oldcount ] */
    duk_pop(ctx, 3);

    return ptr;
}
```

And unpin, using a `void *`:

```c
/* Unpin value using a value from duk_get_heapptr(). */
void duk_unpin_heapptr(duk_context *ctx, void *ptr) {
    duk_uint_t pin_count;

    if (ptr == NULL) {
        return;
    }

    duk_push_heap_stash(ctx);
    duk_push_pointer(ctx, ptr);  /* [ ... stash ptr ] */
    (void) duk_get_prop(ctx, -2);  /* -> [ ... stash oldcount ] */
    pin_count = duk_to_uint(ctx, -1);
    DUK_ASSERT(pin_count >= 1);  /* if untrue, all bets are off */

    duk_push_pointer(ctx, ptr);  /* -> [ ... stash oldcount ptr ] */
    if (pin_count == 1) {
        duk_del_prop(ctx, -3);  /* -> delete stash[ptr]; -> [ ... stash oldcount ] */
    } else {
        duk_push_uint(ctx, pin_count - 1);  /* -> [ ... stash oldcount ptr newcount ] */
        duk_put_prop(ctx, -4);  /* stash[ptr]--; -> [ ... stash oldcount ] */
    }
    duk_pop(ctx, 2);
}
```

This should in a pretty straightforward manner; perhaps the backing object shouldn't be the heap stash directly but some other object (which Duktape internals can maintain a direct reference to). Also Duktape internals could do the necessary operations a bit more directly.

This wouldn't provide any way of tracing leaked pin values etc, so I'm not very eager to merge it into master as is.

**Comment by fatcerberus** [comment]
I should note that this isn't all theoretical from my point of view.  The reason I got more interested in pinning recently is that I'm trying to write a JS abstraction layer for miniSphere in order to make it easier to experiment with different JS engines, so I have a "JS value" struct which stores a reference to an Ecmascript value:
https://github.com/fatcerberus/minisphere/blob/chakra-js/src/shared/jsal.c#L42-L46

Duktape's primary mechanism for passing values is indirectly ("indirectly" from the point of view of C code), via the value stack.  Chakra instead uses `JsRef` variables to allow you to pass object references within C code directly.  In order to cover both styles with a single abstraction, I have to either A) emulate the value stack for direct-drive engines or B) get a heapptr from Duktape and pin it.

I should note that I already have a pinning helper that functions like what you described:
https://github.com/fatcerberus/minisphere/blob/v4.8.4/src/minisphere/utility.c#L212-L255

It's more complex than your code because it's not enough just to key against the pointer for a refcount--if you want to keep it safe from GC, you also have to store the reference itself as well.  It wasn't a trivial function to write and took me a bit to get all the bugs out :)

**Comment by svaarala** [comment]
Oops, you're right, my code example should use one map to track the pin count, and another to pin the object itself.

**Comment by fatcerberus** [comment]
...why didn't I think to use two separate maps?  It probably would have made the code shorter without the nested objects. 😝

**Comment by svaarala** [comment]
> The reason I got more interested in pinning recently is that I'm trying to write a JS abstraction layer for miniSphere [...]

Interesting, let me know how it goes :) I know a few other cases where an application tries to implement an abstraction layer; it's not a trivial issue because of the differences in the basic API approach. A related approach is to use automatic binding generators but they have their own downsides.

For such an abstraction to work well with both direct references and a value stack, it'd be nice to have pinning/unpinning as directly supported operations so they worked as quickly and memory efficiently as possible. The example above, and other utility approaches, are useful for occasional pinning but not really ideal if a lot of values need to be pinned and unpinned (e.g. if all arguments are pinned for every call).

The main problem in implementing direct support for a minimal stacking pin count is that it adds another field to every object which is not desirable by default. But it's also not desirable for the pinning API calls to be non-functional by default. So I guess an ideal approach would use some indirect approach to pinning (e.g. some internal objects tracking the pinned objects) by default, and then provide a config option to make the pin count explicit for each object, when pinning is needed a lot and the memory trade-off is desirable.

**Comment by fatcerberus** [comment]
If pinning becomes part of the API (and therefore heap pointers become a more important concept), I would strongly recommend making `duk_get_heapptr()` return something more specific than a `void*`, to catch mistakes like what happened to me when `duk_context` was aliased to void (I mysteriously corrupted Duktape state by passing a `ctx` into an unrelated function).

**Comment by svaarala** [comment]
Agreed, although I'm more accustomed to void pointers when the contents are "none of your business", the type checking benefit is probably worth it.

Somewhat related, I'm changing duktape.h so that the internal structures are visible when compiling the application. They're of course entirely off limits, but this allows API macros or API inline functions to access internals directly, making call sites smaller and faster. For example, pinning might be just something like `(heaphdr)->pin_count++, (heaphdr)->ref_count++;` but with an inline function to avoid double evaluation of the argument. That's not directly related to your request because you can of course cast back and forth, but it'd make pinning quite fast.

**Comment by fatcerberus** [comment]
One case where I could see pinning being useful in everyday use is symbols.  Right now they are represented as strings so it's not an issue just to use an internal property, but assuming symbols were a unique type, one could do:

```c
duk_push_symbol(ctx, "internalValue");  // like Symbol() function
void* symbol = duk_get_heapptr(ctx);
duk_pin_heapptr(symbol);

/* ... */

duk_push_heapptr(symbol);
duk_push_number(812);
duk_put_prop(ctx);
```
